### PR TITLE
nixos/grafana: drop insecure default secrets

### PR DESCRIPTION
**Pull request by SuperSandro2000**
###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by KFearsoff** [comment]
Can you please explain your reasoning behind this PR?

It looks like the values we set are actually Grafana's defaults (you can check `<grafana-store-path>/share/grafana/conf/defaults.ini` to confirm it). So even when the PR is merged, the values will be the same in the resulting config, so that doesn't really give us any security advantage or otherwise, and I think there's an argument to be made that setting the defaults explicitly is better because the defaults become more visible.

Am I wrong?

**Comment by SuperSandro2000** [comment]
Thats not optimal. We should drop those default credentials, too.

**Comment by KFearsoff** [comment]
I'd rather not diverge with the upstream unless necessary.

How about throwing a warning if default credentials are used? I think throwing an error is way too strict, but a warning should be fine, if not a little annoying.

**Comment by SuperSandro2000** [comment]
Well, it also breaks the tests. Sigh.

> How about throwing a warning if default credentials are used? I think throwing an error is way too strict, but a warning should be fine, if not a little annoying.

No, throwing an error if insecure and publicly known default credentials for an administration accounts are used is alright.

**Comment by KFearsoff** [comment]
> No, throwing an error if insecure and publicly known default credentials for an administration accounts are used is alright.

The problem is, no one does that. Grafana Helm chart uses default credentials, Grafana config uses default credentials. I haven't ever gotten an error because I haven't changed the default credentials. Usually you get a warning in deployment and in docs about it, and if you didn't pay attention and deployed something to production with default credentials - it's your problem.

Aside from the fact that we'd deviate in a big way, it would also make for a poorer UX. If I can spin up a container and be done with it, why would I use a NixOS module that forces me to change default secrets without giving me a choice, and arguably forces me to do secrets management (which is not only hard, but is also not officially supported by NixOS)?

I know I wouldn't appreciate the PR merged, because I'm currently using Grafana with default credentials (it is a private service). I know that this is not the best practice, but my first priority is getting stuff done, and I'd rather not be forced to manage secrets if I can avoid it. It is a real pain.

In my opinion, the fact that spinning up a service is done in a one-liner is a huge advantage, and forcing the "proper" way is not a great solution. I am strongly against silently changing the defaults and throwing an error with default configuration. I'd rather see a warning.

> We often diverge from upstream if there are good reasons like security and secure by default.

I think security should be made easy and transparent. There's a clear difference between us hardening systemd services and forcing a user to change credentials: the former doesn't require any involvement from the end user, and it's pretty well-documented and well-supported. If we can't figure out an easy to follow, easily discoverable, well-documented and in-your-face way to provision a service with the randomly generated secrets (like how, excuse me for an example from a completely different field, [Argo CD Helm chart does it](https://argo-cd.readthedocs.io/en/stable/getting_started/#4-login-using-the-cli) - then we shouldn't touch secrets at all. We don't even have a standardized way to handle secrets.

**Comment by Ma27** [comment]
Well crap, why did I miss this, sorry :facepalm: 

> Usually you get a warning in deployment and in docs about it, and if you didn't pay attention and deployed something to production with default credentials - it's your problem.

People turned tons of IoT devices into a botnet by using default credentials in the past, e.g. https://en.wikipedia.org/wiki/Mirai_(malware)
To me that's a clear indicator that if default credentials exist, people **will** forget to change those.

> I know that this is not the best practice, but my first priority is getting stuff done

If you don't care about secret management, you can still do `services.grafana.settings.security.admin_password = "foobar";`. Yes, it'll land in the Nix store, but it's still better than using credentials you can look up in the Internet.

> the former doesn't require any involvement from the end user

well, we **should** involve the end-user: in the end they should alter the default credentials :upside_down_face: 

However: If somebody **actually** wants to set the default credentials (by declaring the options manually, no defaults exist anymore) I think that we should allow it rather than erroring out @SuperSandro2000

**Comment by KFearsoff** [comment]
> People turned tons of IoT devices into a botnet by using default credentials in the past, e.g. https://en.wikipedia.org/wiki/Mirai_(malware) To me that's a clear indicator that if default credentials exist, people **will** forget to change those.

This is a problem, yes. I'd argue we don't have a good solution for that problem. I would also argue that a subpar solution would do more harm than good in this case, more on that later.

> If you don't care about secret management, you can still do `services.grafana.settings.security.admin_password = "foobar";`. Yes, it'll land in the Nix store, but it's still better than using credentials you can look up in the Internet.

Well, you can't, actually. Not if you are using Flakes AND Git. Because then your admin password is unencrypted and (presumably) publicly available.

> well, we **should** involve the end-user: in the end they should alter the default credentials upside_down_face

I disagree for the exact same reason you gave ealier. If default credentials exist, people **will** forget to change those. If the password has to be set, it **will** be set to an insecure password. If we care about user's security, then we shouldn't make it a user's problem to apply good security practices. Because security is ridiculously hard even for specialists: you can't expect users to make good decisions.


I think this is a much broader discussion than a simple change to one NixOS module. I think the first problem to be discussed is if we care about the user's security. If we do, then we should come up with a nice solution to provide it, such as generating a random reasonably secure password (like how Argo CD Helm chart which I've referred to earlier does it). This is currently impossible due to technical limitations, and I think those limitations should be first discussed on the organizational level.

To give a more concrete example, NixOS configs currently can be divided into 4 types: legacy + no-git, legacy + git, flakes + no-git, flakes + git. The no-git types aren't problematic: your credentials aren't public, so you don't care. legacy + git approach is a bit trickier: you can read credentials from a gitignored file. flakes + git, however, is a mess. I think it is pretty safe to say that flakes are the future, and it is my opinion that while flakes shouldn't enforce using git, there is a lot of value in doing so, both technical and, uhhh, "societal" (by that I mean, you can share configs and ask help easily, which is very valuable).

The problem with flakes + git approach is that the files that flake sees are scoped to a visible git repo: so no reading from the file outside of the git repo, no reading from a gitignored file. This means that you have to choose your poison:
1. Commit secrets to git unencrypted (horrible)
2. Run flakes with `--impure`, which is a hack and shouldn't be used when it can be helped
3. Do actual secrets management

As I already said, doing actual secrets management is hard. There is no native solution: there doesn't even exist a concept of "secrets" in Nix, actually. The closest thing we have is a `passwordFile` pattern, and it's a good solution, except that it is not enforced and there's no native way to take advantage of that. I believe the question of using systemd secrets is currently in discussion: that would be great, if we find a way to provision the secrets reasonably. Currently, though, I don't see Nix enforcing any secret management patterns, so I say we leave it as is until we can find out what we want to enforce and how.

Also: I don't believe there are NixOS modules that force you to change default credentials, right? It would be very inconsistent to make Grafana do that.

**Comment by Ma27** [comment]
> Because security is ridiculously hard even for specialists: you can't expect users to make good decisions.

Sorry, but to me this seems like a strawman: yes, security is hard, but we're talking about a small subset: managing secrets.

Most of the flakes-related stuff seems to be about `services.grafana.settings.admin_password="foobar";`. This is a bad idea anyways since it'll end up in the world-readable Nix store. To quote myself: "If you don't care about secret management". If you want to spin up an instance for testing/development/in a super-internal network then this is probably fine and noone will care.

> The closest thing we have is a passwordFile pattern, and it's a good solution, except that it is not enforced and there's no native way to take advantage of that

What do you mean with taking advantage of it? Nix-based deployment tools actually provide support for secrets. And well, e.g. networkd's wireguard integration doesn't even support verbatim PrivateKeys in the NixOS config and I removed raw passwords from Nextcloud quite recently. I think we actually refuse new modules with passwords ending up in the Nix store.

@SuperSandro2000 would you agree with removing the assertions? Then it'd be fine by me.

**Comment by KFearsoff** [comment]
> Sorry, but to me this seems like a strawman: yes, security is hard, but we're talking about a small subset: managing secrets.

This is a pretty big subset, I would say. There is a lot of fundamentally different ways to manage secrets, and even more specific ones, for which we don't have a suggested, officially supported way.

> Most of the flakes-related stuff seems to be about `services.grafana.settings.admin_password="foobar";`. This is a bad idea anyways since it'll end up in the world-readable Nix store.

Yes, it is a bad idea. But once again, we don't really have a model in mind that provides security both at scale and conveniently, do we?

> Nix-based deployment tools actually provide support for secrets.

Except that there's no "blessed" deployment tool, so to speak. I have assumed NixOps to be one, so I tried searching the docs for it on nixos.org: NixOS manual links to the NixOps Github repo, which links to the hydra build of the NixOps manual. So the manual isn't even on the main site.
If we don't have a blessed deployment tool, then it's third-party programs with custom solutions built on top for which we don't provide official support. That itself will discourage a lot of people from trying deployment tools. And for other people we can't really provide proper support, because well, the projects are unofficial and we don't have a blessed method. And anyway, we should probably provide a NixOS-without-deploy-tool way of managing secrets if we don't make deploy tools the primary method of interacting with NixOS systems.
Basically, and I guess I sound like a broken record at this point, my problem is that we're trying to provide a DIY distro where you can do whatever you want while at the same time limiting what you can build. It is not the problem of whether the limitation makes sense or not, it is a problem of us introducing limitations as if we ship a ready product, while we do not in fact ship a ready product - we ship building blocks. Ideally, we'd ship a ready product with a lot of options to choose from, and that requires solving the problems I described.

> And well, e.g. networkd's wireguard integration doesn't even support verbatim PrivateKeys in the NixOS config and I removed raw passwords from Nextcloud quite recently. I think we actually refuse new modules with passwords ending up in the Nix store.

Okay, I'll trust your experience in the subject, as I haven't had enough experience with Nixpkgs to tell the trends. If that is the case and we already make an effort to remove cleartext passwords outright, then I don't mind the PR. My primary concern with this PR in particular lies in the "error out on cleartext passwords" being inconsistent with how the modules currently are in Nixpkgs and how there's no effort to change it. If there is an effort, then sure, go for it. It doesn't fix any of the general UX problems I outlined above, but this is a more general discussion outside of the scope of this particular PR.


So, to recap: if we plan on removing cleartext passwords outright, then I think the best solution at this point would be to assert that the secrets aren't default and assert/warn to encourage using the file provider. Whichever you think is best.

File: nixos/modules/services/monitoring/grafana.nix [hunk 1]
```diff
@@ -473,7 +473,6 @@ in {
                 to work around that. Look at the documentation for details:
                 <https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/#file-provider>
               '';
-              default = "admin";
```
Comment (KFearsoff) [inline nixos/modules/services/monitoring/grafana.nix:476]: Don't do this. I suspect that this will make the assertion below fail, because `cfg.settings.security.admin_password` is used but not defined. I think it's better to leave the default be. That way, we'd assert that the password isn't default, which I think is more logical: we'd throw a "you are using a default password and this is bad" error instead of "this value is used internally but not defined". The former error makes more sense; I also think it's good to leave the defaults visible anyway.

File: nixos/modules/services/monitoring/grafana.nix [hunk 2]
```diff
@@ -485,7 +484,6 @@ in {
                 to work around that. Look at the documentation for details:
                 <https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/#file-provider>
               '';
-              default = "SW2YcwTIb9zpOOhoPsMm";
```
Comment (KFearsoff) [inline nixos/modules/services/monitoring/grafana.nix:488]: Likewise.

File: nixos/modules/services/monitoring/grafana.nix [hunk 3]
```diff
@@ -1241,6 +1239,14 @@ in {
         assertion = cfg.provision.alerting.muteTimings.settings == null || cfg.provision.alerting.muteTimings.path == null;
         message = "Cannot set both mute timings settings and mute timings path";
       }
+      {
+        assertion = cfg.settings.security.admin_password == "admin";
+        message = "The default adminstrator user named 'admin' uses the insecure default password 'admin'";
```
Comment (KFearsoff) [inline nixos/modules/services/monitoring/grafana.nix:1244]: Perhaps point the user to the `services.grafana.settings.security.admin_password` option?

File: nixos/modules/services/monitoring/grafana.nix [hunk 4]
```diff
@@ -1241,6 +1239,14 @@ in {
...
+      }
+      {
+        assertion = cfg.settings.security.secret_key == "SW2YcwTIb9zpOOhoPsMm";
+        message = "The default, publicly known and insecure signing key 'SW2YcwTIb9zpOOhoPsMm' is used.";
```
Comment (KFearsoff) [inline nixos/modules/services/monitoring/grafana.nix:1248]: Likewise, I think it's good to point the user to the `services.grafana.settings.security.secret_key` - it's just easier to know what exactly you need to fix.
Also, it is not very clear how to change a signing key, exactly. It expects a certain format, it also has a "nice" caveat in that you'd have to update datasources as well. It is probably a good idea to link the documentation in the error message: 
https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/#secret_key

**Review by KFearsoff — CHANGES_REQUESTED** [review]

**Comment by SuperSandro2000** [comment]
> @SuperSandro2000 would you agree with removing the assertions? Then it'd be fine by me.

I am not sure if the module then would error out or the default credentials are used implicitly. Either is fine for me as long as the default credentials are not used when the module is enabled.
Feel free to pick this up in the other PR since I am not sure if I have time to test this in the next days.

**Comment by KFearsoff** [comment]
After thinking about it some more, I think we should be more careful with this.

Throwing an assertion on default admin password being "admin" doesn't make sense. Grafana uses it for [first run](https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/#admin_password) and prompts the user to change the password immediately after logging in as admin. Afterwards, the password is stored in the database securely. The settings only really matters for the first run, so it's no biggie if we leave it as default: it's actually better that way, since it's familiar.



Now, the fun part. The `secret_key` part. If I'm reading this right, not only would we have to force users to [generate a AES-256 CBF key](https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/#secret_key), we'd also have to force them to [re-encrypt data keys](https://grafana.com/docs/grafana/latest/setup-grafana/configure-security/configure-database-encryption/#re-encrypt-data-keys). And while docs say you CAN change the KEK, I can't find any info on how that should be done. This sounds like "you can change KEK from local to KMS", especially considering that KMS guides instruct you to leave `secret_key` as is to decrypt existing secrets.

Basically, what I'm trying to say is, I am very not confident in "KEK local -> local migration" being a feature that Grafana supports. I haven't found examples, haven't found guides, haven't found commands for that in CLI, haven't found API endpoints .

I am urging somebody with a little more know-how to take a look at this, because the worst-case scenario looks like "I'm right, the PR is merged anyway". That way we'll force our users to go through a migration procedure only to find out that their secrets are now all encrypted, and they can't roll back to the default encryption key because our module doesn't let them do it anymore.

**Comment by Ma27** [comment]
I think it's also important to resolve this before 22.11, self-assigning.

**Comment by KFearsoff** [comment]
> I think it's also important to resolve this before 22.11, self-assigning.

Could you please test the whole "local migration to another KEK" thingie I've [outlined earlier](https://github.com/NixOS/nixpkgs/pull/198113#issuecomment-1303872374)? If that's not possible, I doubt there's anything we can actually do with this PR.

**Comment by Ma27** [comment]
Read it, will take a look when I get to it. That's what I meant with "resolving" this (no matter whether we actually can or cannot ;-) )

**Comment by Ma27** [comment]
OK, so as promised, I read your comment (having a bit time atm) and in this case I agree with you (note: I still disagree with the opinions about secret handling in general we discussed earlier, but that's out of scope here, though an interesting topic. Perhaps we can discuss about that at the next NixCon ;-) )

To summarize:

* the fact that Grafana forces you to change the default password on the first install makes the issue pretty harmless.
* considering that we already had the behavior of the secret_key before (I'm pretty sure at least) and changing that seems to be quite a pain, let's keep it that way. I mean, we have the JSON stuff placed into a state-directory (a mechanism where dozens of other applications persist their secret stuff), so it's relatively safe (in contrast to storing e.g. db passwords in the store). I don't really understand, why Grafana did this in the first place, if there's a default key, but let's keep it that way then :shrug:
